fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -2548,4 +2641,220 @@ public boolean isContinueIfWriteOnly() {
             return continueIfWriteOnly;
         }
     }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndexMarkBuilt(Subspace subspace) {
+        // zz 2.1 - mark built
+        // Grand Finale - after fully building the index, remove all missing ranges in one gulp
+        // Note that we do not assume srcIndex has the same order as the record store, therefore did not try to handle
+        // missing spaces (for now..)
+        return getRunner().runAsync(context ->
+        {
+            RangeSet rangeSet = new RangeSet(subspace.subspace(Tuple.from(FDBRecordStore.INDEX_RANGE_SPACE_KEY, index.getSubspaceKey())));
+            TransactionContext tc = context.ensureActive();
+            return rangeSet.insertRange(tc, null, null).thenApply(bignore -> null);
+        });
+    }
+
+    private void buildIndexFromIndexThrowEx() {
+        throw new MetaDataException(IndexFromIndexRequest.BAD_SRC_MESSAGE);
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndexReadableOrThrow(FDBRecordStore store, String srcIndex) {
+        return store.getRecordStoreStateAsync()
+                .thenCompose(state ->
+                {
+                    if (!state.isReadable(srcIndex)) {
+                        buildIndexFromIndexThrowEx();
+                    }
+                    return AsyncUtil.DONE;
+                });
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndexAsync() {
+        // zz 0 - sanity & trigger
+        return getRunner().runAsync(this::openRecordStore)
+                .thenCompose(store ->
+                {
+                    // first verify that both src and tgt are of a single, similar, type
+                    final RecordMetaData metaData = store.getRecordMetaData();
+                    final Collection<RecordType> srcRecordTypes = metaData.recordTypesForIndex(index);
+
+                    if ( syntheticIndex ||
+                            recordTypes.size() != 1 ||
+                            srcRecordTypes.size() != 1 ||
+                            recordTypes.toArray()[0] != srcRecordTypes.toArray()[0]) {
```

Comment:
"Cannot resolve method 'get' in 'Collection'" 
How about `recordTypes.equals(srcRecordTypes)`?

Reply:
This `equals` doesn't work. I suppose it's safer to compare by pointer.

Reply:
As to why `.get()` doesn't work, I guess that's because record types is a `Collection` rather than `List`, which makes some amount of sense...

Why doesn't `.equals` work? That seems like what we "should" be using, in some sense.

Reply:
Shouldn't we implement an `equals` method in RecordType to make it work as expected?